Fix rollback of TxMeta and Certs

Anviking opened the pull request:
# Issue Number

#1575 

# Overview

- [x] I have constructed a regression test using the DB and a mock chain.
- [x] I tweaked the `rollbackTo` code to clear from `nearestPoint` instead of `point`. I.e. the actual rollback point instead of the requested rollback point.
- [x] Improved state machine counterexamples slightly

# Comments

- We should perhaps think about why no higher-level property caught this bug previously.

<!-- Additional comments or screenshots to attach if any -->

<!-- 
Don't forget to:

 ✓ Self-review your changes to make sure nothing unexpected slipped through
 ✓ Assign yourself to the PR
 ✓ Assign one or several reviewer(s)
 ✓ Once created, link this PR to its corresponding ticket
 ✓ Assign the PR to a corresponding milestone
 ✓ Acknowledge any changes required to the Wiki
-->

Anviking committed:
Add failing test for rollback of TxMeta

If we have:

Checkpoints: cpA              | cpB
TxMeta:          m1           |
Certs:                     c1 |
--------------------------------------> Increasing slotId
                              |
                              |
                              + p, rollback to here

Then we should rollback to cpA and remember to delete
m1 and c1!

Checkpoints: cpA
TxMeta:
Certs:
--------------------------------------> Increasing slotId

This commit adds a test that fails with:

test/unit/Cardano/Wallet/DB/SqliteSpec.hs:571:17:
  1) Cardano.Wallet.DB.Sqlite, Check db reading/writing from/to file and cleaning, Golden rollback scenarios, (Regression test #1575) - TxMetas and checkpoints should rollback to the same place
       expected: []
        but got: [(Outgoing,2),(Incoming,4)]

  To rerun use: --match "/Cardano.Wallet.DB.Sqlite/Check db reading/writing from/to file and cleaning/Golden rollback scenarios/(Regression test #1575) - TxMetas and checkpoints should rollback to the same place/"

Randomized with seed 3967246

Anviking committed:
Fix rollback of TxMeta

Fixes test from previous commit.

Anviking commented on lib/core/test/unit/Cardano/Wallet/DB/SqliteSpec.hs:
```diff
@@ -10,7 +10,7 @@
 {-# LANGUAGE TypeApplications #-}
 
-{-# OPTIONS_GHC -fno-warn-orphans #-}
+{-# OPTIONS_GHC -fno-warn-orphans -fno-warn-unused-imports #-}
```
Oops

Anviking commented on lib/core/src/Cardano/Wallet/DB/Sqlite.hs:
```diff
@@ -505,21 +505,21 @@ newDBLayer trace defaultFieldValues mDatabaseFile = do
                 Just nearestPoint -> do
                     deleteCheckpoints wid
-                        [ CheckpointSlot >. point
+                        [ CheckpointSlot >. nearestPoint
```
🤔 Should work, but the change should also not be necessary.

Anviking replied:
Keeping it for consistency.

Anviking commented on lib/core/src/Cardano/Wallet/DB/Sqlite.hs:
```diff
@@ -505,21 +505,21 @@ newDBLayer trace defaultFieldValues mDatabaseFile = do
+                        [ CheckpointSlot >. nearestPoint
                         ]
                     deleteDelegationCertificates wid
                         [ CertSlot >. point
```
Should also be `nearestPoint`

KtorZ commented on lib/core/test/unit/Cardano/Wallet/DB/SqliteSpec.hs:
```diff
@@ -471,6 +484,88 @@ fileModeSpec =  do
+                    rollbackTo testPk (SlotId 2 0)
+                Just cp <- atomically $ readCheckpoint testPk
+
+                -- We end up at genesis.
```
That comment is superfluous. Comments should give detail about "why" , not "what". What is written just below and already perfectly readable :ok_hand:

Anviking replied:
> That comment is superfluous

Yes, perhaps. I totally agree in general.

My reasoning might have been: There is a lot of boilerplate here and "What"-comments highlight the important stuff.

KtorZ replied:
Functions should be used then, they are clearer than comments, and re-usable.

Anviking replied:
755065b

KtorZ commented on lib/core/test/unit/Cardano/Wallet/DB/SqliteSpec.hs:
```diff
@@ -471,6 +484,88 @@ fileModeSpec =  do
+                -- (Outgoing, 2) should /not/ be a "InLedger" transaction!
+                --
+                -- Whether is is now pending, or removed completely is not the
+                -- concern of this test.
```
Same as above, the comment is superfluous.

KtorZ commented on lib/core/test/unit/Cardano/Wallet/DB/SqliteSpec.hs:
```diff
@@ -471,6 +484,88 @@ fileModeSpec =  do
+                                { slotId = slotA { epochNumber = (epochNumber slotA) + 1}
+                                -- Increment blockHeight by steps greater than k
+                                -- Such that old checkpoints are always pruned.
+                                , blockHeight = Quantity $ bhA + 5000
```
I am not sure what the pruning has to do with this test and why it matters to introduce this extra complexity in the test?

Anviking replied:
Hm… indeed

Anviking replied:
No, this is essential. Without pruning there is no problem.

If a checkpoint exists for every slot, then `nearestPoint` and `requestedPoint` are going to be equal, and the difference does not matter (Also see diagram in #1575)

KtorZ replied:
But you don't have to insert a checkpoint at every slot for the test in the first place?

Anviking replied:
I wanted to include [`prune`](https://github.com/input-output-hk/cardano-wallet/blob/7e0aa90186462721397b22a3ca90967862716fbb/lib/core/src/Cardano/Wallet.hs#L767) similar to `restoreBlocks`, to convince myself that the bug could happen in reality.

But, yes, we could also specify which checkpoints to write manually.

KtorZ reviewed (commented):
Looks good, although I wonder why not writing a simple property here using arbitrary data rather using mock and fake and dummy everywhere :thinking:  ?

Also, seems like `prop_rollbackTxHistory`, for a good reason it seems. The test isn't actually taking care of any consistency consideration between checkpoints and history. 

Also, the database model needs to be updated accordingly, hence the state-machine failures.

Anviking commented on lib/core/test/unit/Cardano/Wallet/DB/Properties.hs:
```diff
@@ -726,10 +726,13 @@ prop_rollbackTxHistory db@DBLayer{..} (InitialCheckpoint cp0) (GenTxHistory txs0
         txs <- run $ atomically $ readTxHistory wid Descending wholeRange Nothing
-        monitor $ counterexample $ "\nTx history after rollback: \n" <> fmt txs
+
+        monitor $ counterexample $ "\n" <> "Actual Rollback Point:\n" <> (pretty point)
```
This is weird. 
```
         Requested Rollback point:
         79.0

         Actual Rollback Point:
         0.0

         Original tx history:
         -  9eb44fb2 14.6 in ledger outgoing 2

         New tx history:
         -  9eb44fb2 14.6 in ledger outgoing 2
         Outgoing txs are reschuled ✗

  To rerun use: --match "/Cardano.Wallet.DB.MVar/MVar/rollback/Correctly re-construct tx history on rollbacks/"
```

Through experimentation it seems the `TxMetaSlot >. nearestPoint` predicate is ineffective. (If removed all outgoing become pending, all incoming are removed). But still: wat?

Anviking replied:
Not sure what fixed it 🤷‍♂️

Anviking committed:
Make (Wallet s) a plain ADT to derive Generic

Generic is useful for ToExpr instances for pretty
test failures.

Anviking committed:
Adjust prop_rollbackTxHistory and improve counterexamples

Failure:
             Wallet ID:
             764f9aff...e868af41

             Wallet Metadata:
             charmander, created at 1903-10-09 18:30:21 UTC, delegating to ?

             Requested Rollback point:
             92.4

             Actual Rollback Point:
             0.0

             Original tx history:
             -  a4460aa7 31.1 in ledger incoming 0

             New tx history:
             -  a4460aa7 31.1 in ledger incoming 0
             Outgoing txs are reschuled ?
             All other txs are still known ?

Anviking committed:
Add more shrinking to State Machine Tests

Without this commit,
and with out the recent model fix,
running

$ stack test cardano-wallet-core --ta '--match "/Cardano.Wallet.DB.Sqlite/Sqlite State machine (RndState)/Sequential state machine tests/"'

I get a PostconditionFailed message of
==> 57k characters.

With this commit the failure is only
==> 7k characters.

Anviking committed:
Derive ToExpr instances using generic

This makes the State Machine Tests highlight the
changes to the model after each step.

Anviking committed:
regenerate nix

KtorZ commented on lib/core/test/unit/Cardano/Wallet/DB/Arbitrary.hs:
```diff
@@ -413,6 +414,14 @@ instance Arbitrary UTxO where
+
+    shrink (Range from to) =
+        [Range from' to | from' <- shrink from]
+        ++ [Range from to' | to' <- shrink to]
```
Are we sure that `to'` is necessarily bigger than `from` here ? For `from'`, it should shrink to a value smaller than `from` so there's no obvious problem, but for `to'` :shrug:

Anviking committed:
fixup: hlint

Anviking committed:
fixup: make sure shrunk ranges are valid

KtorZ reviewed (approved):
LGTM. I wonder how this previously property got fixed though...
